### PR TITLE
Configurable log rotation sizes via new agent-logfile-* settings

### PR DESCRIPTION
**Pull request by benhoyt**
Allow configuring various log file sizes via controller config. This adds two new controller config keys, `agent-logfile-max-size` and `agent-logfile-max-backups`, and makes the following logfile rotations configurable:

* /var/log/juju/logsink.log (apiserver logsink)
* /var/log/juju/machine-0.log (machine agent), on controller machines
* /var/log/juju/machine-0.log (machine agent), on unit machines
* /var/log/juju/unit-mysql-0.log (unit agent)

The model log /var/log/juju/models/test-123456.log is already configurable via `model-logfile-max-size` and `model-logfile-max-backups`.

The PR also includes a drive-by fix so that `juju bootstrap -h` shows all the config keys, not just the updatable ones (and `juju controller-config -h` still shows only the updatable ones).

## QA steps

First you need some way to generate tons of logs, as the minimum logfile rotation size is 1MB. I added the following code to just after the relevant lumberjack config setup, to add 50,000 bytes of logs every 3 seconds:

```go
    go func() {
      for { 
        time.Sleep(3 * time.Second)
        line := []byte(fmt.Sprintf("%s: TODO2 The quick brown fox jumps over the lazy dog. Padding this out to 100!!!\n", time.Now().Format(time.RFC3339)))
        lines := bytes.Repeat(line, 500) // will take about 60s to exceed 1MB
        _, err := ljLogger.Write(lines)
        if err != nil {
          a.logger.Errorf("NewFileWriter: TODO: error writing to ljLogger: %v", err)
        }
      }
    }()
```

Here are the commands I use for manual testing the various log files:

```
# /var/log/juju/logsink.log (apiserver.logsink.NewFileWriter)
$ juju bootstrap localhost --config agent-logfile-max-size=1MB
$ lxc exec juju-a10f27-0 -- ls -al /var/log/juju
...
-rw-r----- 1 syslog adm      3580 Jan 11 02:52 logsink-2022-01-11T02-52-26.225.log.gz
-rw-r----- 1 syslog adm      3581 Jan 11 02:53 logsink-2022-01-11T02-53-26.245.log.gz
-rw-r----- 1 syslog adm    150000 Jan 11 02:53 logsink.log

# /var/log/juju/machine-0.log (cmd/jujud/agent:machineAgentCmd.Init)
# on controller machine
$ juju bootstrap localhost --config agent-logfile-max-size=1MB
$ lxc exec juju-6bf629-0 -- ls -l /var/log/juju
...
-rw-r----- 1 syslog adm   3577 Jan 12 02:01 machine-0-2022-01-12T02-01-07.995.log.gz
-rw-r----- 1 syslog adm   3578 Jan 12 02:02 machine-0-2022-01-12T02-02-08.011.log.gz
-rw-r----- 1 syslog adm 600000 Jan 12 02:02 machine-0.log
$ lxc exec juju-6bf629-0 -- cat /var/lib/juju/agents/machine-0/agent.conf | grep logfile
agent-logfile-max-size: 1
agent-logfile-max-backups: 2

# on unit machine (eg: mysql)
$ lxc exec juju-f667b0-0 -- ls -al /var/log/juju
...
-rw-r----- 1 syslog adm    1110666 Jan 11 03:12 machine-0.log

# models/test-123456.log (cmd/jujud/agent:MachineAgent.startModelWorkers)
# this one was already configurable
$ juju bootstrap localhost --config model-logfile-max-size=1MB
$ lxc exec juju-b3795c-0 -- ls -al /var/log/juju/models
total 508
...
-rw-r----- 1 syslog adm   3579 Jan 11 03:31 admin-default-4575e0-2022-01-11T03-31-08.880.log.gz
-rw-r----- 1 syslog adm   3580 Jan 11 03:32 admin-default-4575e0-2022-01-11T03-32-08.897.log.gz
-rw-r----- 1 syslog adm 500000 Jan 11 03:32 admin-default-4575e0.log

# unit-mysql-0.log (worker/deployer:UnitAgent.initLogging)
$ juju bootstrap localhost --config agent-logfile-max-size=1MB
$ juju deploy mysql
$ lxc exec juju-db4775-0 -- ls -al /var/log/juju
...
-rw-r----- 1 syslog adm    1301665 Jan 11 23:05 unit-mysql-0.log
```

## Documentation changes

TODO

## Bug reference

https://bugs.launchpad.net/juju/+bug/1955750

### CODE REVIEW

File: cmd/juju/commands/bootstrap.go [hunk 1]
```diff
@@ -284,7 +284,7 @@ func (c *bootstrapCommand) configDetails() map[string]interface{} {
 	if modelCgf, err := cmdmodel.ConfigDetails(); err == nil {
 		addAll(modelCgf)
 	}
-	if controllerCgf, err := cmdcontroller.ConfigDetails(); err == nil {
+	if controllerCgf, err := cmdcontroller.ConfigDetailsAll(); err == nil {
```
Comment (benhoyt) [inline cmd/juju/commands/bootstrap.go:287]: This is a drive-by fix so that `juju bootstrap -h` shows all the config keys, not just the updatable ones (and `juju controller-config -h` still shows only the updatable ones).

File: cmd/juju/gui/gui.go [hunk 2]
```diff
@@ -175,7 +175,9 @@ func (c *guiCommand) Run(ctx *cmd.Context) error {
 		controllerName,
 	)
 
-	signalCh := make(chan os.Signal)
+	// Buffer size of 1 avoids race condition if signal is sent between
```
Comment (benhoyt) [inline cmd/juju/gui/gui.go:178]: A drive-by fix to get the Juju repo "go vet" clean (at least `go vet -composites=false`).

File: controller/config.go [hunk 3]
```diff
@@ -300,6 +308,14 @@ const (
 	// batches we will process. (deprecated)
 	DefaultMaxPruneTxnPasses = 100
 
+	// DefaultAgentLogfileMaxSize is the maximum file size in MB of each
+	// agent/controller log file.
+	DefaultAgentLogfileMaxSize = 300
```
Comment (wallyworld) [inline controller/config.go:313]: I think the default here could be a little lower, like 100. The primary source of logs is still debug-log which goes straight to mongo.

Comment (benhoyt) [reply controller/config.go]: Done, thanks.

**Comment by benhoyt** [comment]
!!build!!

File: worker/provisioner/provisioner_task.go [hunk 4]
```diff
@@ -782,6 +782,12 @@ func (task *provisionerTask) constructInstanceConfig(
 		return nil, errors.Trace(err)
 	}
 
+	instanceConfig.Controller = &instancecfg.ControllerConfig{}
```
Comment (wallyworld) [inline worker/provisioner/provisioner_task.go:785]: I think this is ok to move here.
I can't think of a case where `len(pInfo.Jobs) > 0` will not be true - a node normally has at the least the HostUnits job defined. It would be interesting to explore if we need the jobs check at all.

**Review by wallyworld — APPROVED** [review]
QA instructions are most excellent.

File: worker/provisioner/provisioner_task.go (hunk 4 above)
Comment (benhoyt) [reply worker/provisioner/provisioner_task.go:785]: Yeah, you're right. It's probably not necessary here, unless some code downstream is checking `jobs == nil` vs `len(jobs) == 0` and making a distinction. Going to leave as is for now just in case.

**Comment by benhoyt** [comment]
!!build!!

**Comment by hpidcock** [comment]
!!build!!

**Comment by benhoyt** [comment]
!!build!!

**Comment by benhoyt** [comment]
$$merge$$

**Comment by benhoyt** [comment]
$$merge$$

**Comment by benhoyt** [comment]
$$merge$$

**Comment by hpidcock** [comment]
$$merge$$
